linkd-lib/src/args.rs

```diff
@@ -0,0 +1,298 @@
+            Some((peer_id, addr)) => {
+                let peer_id = peer_id
+                    .parse()
+                    .map_err(|e: crypto::peer::conversion::Error| e.to_string())?;
+
+                Ok(Self {
+                    addr: addr.to_string(),
+                    peer_id,
+                })
+            },
+            None => Err("missing peer id".to_string()),
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, StructOpt)]
+pub enum Signer {
+    /// Construct signer from a secret key.
+    Key,
+    /// Connect to ssh-agent for delegated signing.
+    SshAgent,
+}
+
+impl Default for Signer {
+    fn default() -> Self {
+        Self::SshAgent
+    }
+}
+
+impl fmt::Display for Signer {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let ty = match self {
+            Self::Key => "key",
+            Self::SshAgent => "ssh-agent",
+        };
+
+        write!(f, "{}", ty)
+    }
+}
+
+impl FromStr for Signer {
+    type Err = String;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        match input {
+            "key" => Ok(Self::Key),
+            "ssh-agent" => Ok(Self::SshAgent),
+            _ => Err(format!("unsupported signer `{}`", input)),
+        }
+    }
+}
+
+#[derive(Debug, Default, Eq, PartialEq, StructOpt)]
+pub struct KeyArgs {
+    /// Location of the key file on disk.
+    #[structopt(
+        long = "key-file-path",
+        name = "key-file-path",
+        default_value = "",
+        parse(from_str),
+        required_if("signer", "key"),
+        required_if("key-source", "file")
+    )]
+    pub file_path: PathBuf,
+    /// Format of the key input data.
+    #[structopt(
+        long = "key-format",
+        name = "key-format",
+        default_value,
+        required_if("signer", "key")
+    )]
+    pub format: KeyFormat,
+    /// Specifies from which source the secret should be read.
+    #[structopt(
+        long = "key-source",
+        name = "key-source",
+        default_value,
+        required_if("signer", "key")
+    )]
+    pub source: KeySource,
+}
+
+#[derive(Debug, Eq, PartialEq, StructOpt)]
+pub enum KeyFormat {
+    Base64,
+    Binary,
+}
+
+impl Default for KeyFormat {
+    fn default() -> Self {
+        Self::Binary
+    }
+}
+
+impl fmt::Display for KeyFormat {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let source = match self {
+            Self::Base64 => "base64",
+            Self::Binary => "binary",
+        };
+        write!(f, "{}", source)
+    }
+}
+
+impl FromStr for KeyFormat {
+    type Err = String;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        match input {
+            "base64" => Ok(Self::Base64),
+            "binary" => Ok(Self::Binary),
+            _ => Err(format!("unsupported key format `{}`", input)),
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, StructOpt)]
+pub enum KeySource {
+    File,
+    InMemory,
```

Comment:
I was wondering: "whose memory?"

Maybe this should be called `Ephemeral`, `AdHoc`, `Generate`, or somesuch?

Reply:
ac88165